File: Scala/src/main/scala/TextbookTrack/Chapter11/BA11D.scala
package TextbookTrack.Chapter11

object BA11D {
  import TextbookTrack.Utils.IntegerMassTable

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def reverseMassTableMapping(massTable: Map[Char, Int]): Map[Int, Char] = massTable.map(_.swap)

  private def calcPrefixMasses(peptideVector: List[Int]): List[Int] =
    0 :: peptideVector.zipWithIndex.withFilter{ case (item, _) => item == 1 }.map{ case (_, ix) => ix + 1 }

  def restorePeptideFromPeptideVector(peptideVector: List[Int], massTable: Map[Char, Int]): Option[String] = {
    val prefixMasses: List[Int] = calcPrefixMasses(peptideVector)
    val inverseMapping: Map[Int, Char] = reverseMassTableMapping(massTable)
    val peptide: List[Option[Char]] =
      (for { List(mass1, mass2) <- prefixMasses.sliding(2) } yield inverseMapping.get(mass2 - mass1)).toList
    if (peptide.exists(_.isEmpty)) None else Some(peptide.flatten.mkString)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val peptideVector: List[Int] = convertToIntList(reader.next())
    val result: Option[String] = restorePeptideFromPeptideVector(peptideVector, IntegerMassTable)
    result match {
      case Some(peptide) => println(peptide)
      case None => println("The given vector does not encode a peptide.")
    }
  }
}

File: Python/textbook_track/chapter11/ba11d.py
# Converting Peptide Vector into Peptide Problem
import sys

from textbook_track.resources.utils import read_amino_acid_mass_table


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def reverse_mass_table_mapping(mass_table):
    return {mass: amino_acid for amino_acid, mass in mass_table.items()}


def calc_prefix_masses(peptide_vector):
    prefix_masses = [0]
    for ix, item in enumerate(peptide_vector, start=1):
        if item == 1:
            prefix_masses.append(ix)
    return prefix_masses


def restore_peptide_from_peptide_vector(peptide_vector, mass_table):
    prefix_masses = calc_prefix_masses(peptide_vector)
    inverse_mapping = reverse_mass_table_mapping(mass_table)
    peptide = []
    for mass1, mass2 in zip(prefix_masses, prefix_masses[1:]):
        mass = mass2 - mass1
        if (amino_acid := inverse_mapping.get(mass)) is None:
            return None
        peptide.append(amino_acid)
    return ''.join(peptide)


def main():
    reader = sys.stdin
    peptide_vector = convert_to_intlist(next(reader))
    mass_table = read_amino_acid_mass_table()
    result = restore_peptide_from_peptide_vector(peptide_vector, mass_table)
    print(result)


if __name__ == '__main__':
    main()
